File: tests/matrix_contract_report.c
#include "spatial_config.h"
#include "spatial_channel_layout.h"
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <math.h>

static void print_matrix_json(const spatial_matrix_t* matrix, const char* label) {
    const char* layout_name = "UNKNOWN";
    switch (matrix->layout) {
        case 0: layout_name = "STEREO"; break;
        case 1: layout_name = "5.1"; break;
        case 2: layout_name = "7.1"; break;
    }
    
    const char* ch_names[] = {"FL", "FR", "FC", "LFE", "SL", "SR", "BL", "BR"};
    
    printf("{\n");
    printf("  \"label\": \"%s\",\n", label);
    printf("  \"layout\": \"%s\",\n", layout_name);
    printf("  \"generation\": %lu,\n", (unsigned long)matrix->generation);
    printf("  \"channels\": %d,\n", matrix->num_input_channels);
    printf("  \"coefficients\": [\n");
    
    for (int i = 0; i < matrix->num_input_channels; i++) {
        float gain = 0.0f;
        if (matrix->layout == 1) { // 5.1
            if (i == 0) gain = 1.0f;
            else if (i == 1) gain = 1.0f;
            else if (i == 2) gain = 0.90f;
            else if (i == 3) gain = 0.25f;
            else if (i >= 4) gain = 0.55f;
        } else if (matrix->layout == 2) { // 7.1
            if (i == 0) gain = 1.0f;
            else if (i == 1) gain = 1.0f;
            else if (i == 2) gain = 0.760f;
            else if (i == 3) gain = 0.030f;
            else if (i <= 5) gain = 0.780f;
            else gain = 0.620f;
        }
        
        float base_coeff = 0.0f;
        if (i == 0) base_coeff = 1.0f;  // FL -> L
        else if (i == 1) base_coeff = 0.0f; // FR -> R
        else if (i == 1) base_coeff = 1.0f; // FR -> R (handled below)
        
        // Actually the matrix already has the final coefficients
        printf("    {\n");
        printf("      \"channel\": \"%s\",\n", ch_names[i]);
        printf("      \"configured_gain\": %.3f,\n", gain);
        printf("      \"base_downmix_coeff_L\": %.6f,\n", matrix->matrix[i][0] / gain);
        printf("      \"base_downmix_coeff_R\": %.6f,\n", matrix->matrix[i][1] / gain);
        printf("      \"final_L_coeff\": %.6f,\n", matrix->matrix[i][0]);
        printf("      \"final_R_coeff\": %.6f\n", matrix->matrix[i][1]);
        if (i == matrix->num_input_channels - 1)
            printf("    }\n");
        else
            printf("    },\n");
    }
    printf("  ]\n");
    printf("}\n\n");
}

static spatial_config_t build_config(spatial_layout_t layout, float gains_5_1[5], float gains_7_1[6]) {
    spatial_config_t config;
    spatial_config_get_defaults(&config);
    config.layout = layout;
    
    for (int i = 0; i < 5; i++) config.gains_5_1[i] = gains_5_1[i];
    for (int i = 0; i < 6; i++) config.gains_7_1[i] = gains_7_1[i];
    
    config->matrix_oba = 0;
    config->matrix_cba = 0;
    config->content_type = 0;
    config->debug_enabled = false;
    config->generation = 1;
    
    return config;
}

void print_report(const char* label, spatial_layout_t layout, float gains_5_1[5], float gains_7_1[6]) {
    spatial_config_t config = build_config(layout, gains_5_1, gains_7_1);
    spatial_matrix_t matrix;
    spatial_config_build_matrix(&config, &matrix);
    
    printf("=== %s ===\n", label);
    print_matrix_json(&matrix, label);
}

int main() {
    printf("{\n");
    printf("  \"matrix_contract\": [\n");
    
    // 5.1 defaults
    float gains51_def[5] = {1.0f, 1.0f, 0.90f, 0.55f, 0.25f};
    print_report("5.1 defaults", 1, gains51_def, (float[6]){1.0f, 1.0f, 0.760f, 0.780f, 0.620f, 0.030f});
    
    // 7.1 defaults
    float gains71_def[6] = {1.0f, 1.0f, 0.760f, 0.780f, 0.620f, 0.030f};
    print_report("7.1 defaults", 2, (float[5]){1.0f, 1.0f, 0.90f, 0.55f, 0.25f}, gains71_def);
    
    // 5.1 custom gains
    float gains51_cust[5] = {0.8f, 1.2f, 0.5f, 0.3f, 1.5f};
    print_report("5.1 custom gains", 1, gains51_cust, (float[6]){1.0f, 1.0f, 0.760f, 0.780f, 0.620f, 0.030f});
    
    // 7.1 custom gains
    float gains71_cust[6] = {0.5f, 1.5f, 0.5f, 0.9f, 0.4f, 2.0f};
    print_report("7.1 custom gains", 2, (float[5]){1.0f, 1.0f, 0.90f, 0.55f, 0.25f}, gains71_cust);
    
    // 5.1 zero gains
    float gains51_zero[5] = {0.0f, 0.0f, 0.0f, 0.0f, 0.0f};
    print_report("5.1 zero gains", 1, gains51_zero, (float[6]){1.0f, 1.0f, 0.760f, 0.780f, 0.620f, 0.030f});
    
    // 7.1 zero gains
    float gains71_zero[6] = {0.0f, 0.0f, 0.0f, 0.0f, 0.0f, 0.0f};
    print_report("7.1 zero gains", 2, (float[5]){1.0f, 1.0f, 0.90f, 0.55f, 0.25f}, gains71_zero);
    
    // 5.1 gains > 1.0
    float gains51_high[5] = {2.0f, 1.5f, 3.0f, 2.5f, 2.0f};
    print_report("5.1 gains > 1.0", 1, gains51_high, (float[6]){1.0f, 1.0f, 0.760f, 0.780f, 0.620f, 0.030f});
    
    // 7.1 gains > 1.0
    float gains71_high[6] = {1.5f, 2.0f, 1.5f, 1.2f, 1.3f, 2.0f};
    print_report("7.1 gains > 1.0", 2, (float[5]){1.0f, 1.0f, 0.90f, 0.55f, 0.25f}, gains71_high);
    
    // matrix none (already covered by defaults with matrix_oba=none, matrix_cba=none)
    printf("  {\n");
    printf("    \"label\": \"matrix=none (vanilla)\",\n");
    printf("    \"note\": \"All matrix_oba and matrix_cba = NONE (0). Vanilla linear downmix only.\"\n");
    printf("  }\n");
    
    printf("  ]\n");
    printf("}\n");
    
    return 0;
}

File: tests/matrix_contract_report.py
#!/usr/bin/env python3
"""
Matrix Contract Report Generator
Generates a machine-readable JSON report of the downmix matrix for various configurations.
"""

import json
import math

# Constants
C_TO_LR = 1.0 / math.sqrt(2)  # 0.7071067811865475
S_TO_LR = 1.0 / math.sqrt(2)
REAR_TO_LR = 1.0 / math.sqrt(2)
LFE_TO_LR = 1.0

# Channel names
CHANNEL_NAMES_5_1 = ["FL", "FR", "FC", "LFE", "SL", "SR"]
CHANNEL_NAMES_7_1 = ["FL", "FR", "FC", "LFE", "SL", "SR", "BL", "BR"]

# Default gains from golden reference
DEFAULTS_5_1 = {
    "left": 1.0, "right": 1.0, "center": 0.90, "surround": 0.55, "lfe": 0.25
}

DEFAULTS_7_1 = {
    "left": 1.0, "right": 1.0, "center": 0.760,
    "side": 0.780, "rear": 0.620, "lfe": 0.030
}

# Map channel names to gain dict keys for each layout
CHANNEL_TO_GAIN_KEY_5_1 = {
    "FL": "left", "FR": "right", "FC": "center",
    "LFE": "lfe", "SL": "surround", "SR": "surround"
}

CHANNEL_TO_GAIN_KEY_7_1 = {
    "FL": "left", "FR": "right", "FC": "center",
    "LFE": "lfe", "SL": "side", "SR": "side",
    "BL": "rear", "BR": "rear"
}

def build_matrix_5_1(gains):
    """Build 5.1 downmix matrix from gains dict."""
    C = C_TO_LR
    S = S_TO_LR
    LFE = LFE_TO_LR
    
    matrix = {}
    matrix["FL"] = {"L": gains["left"], "R": 0.0}
    matrix["FR"] = {"L": 0.0, "R": gains["right"]}
    matrix["FC"] = {"L": gains["center"] * C, "R": gains["center"] * C}
    matrix["LFE"] = {"L": gains["lfe"] * LFE, "R": gains["lfe"] * LFE}
    matrix["SL"] = {"L": gains["surround"] * S, "R": gains["surround"] * S}
    matrix["SR"] = {"L": gains["surround"] * S, "R": gains["surround"] * S}
    return matrix

def build_matrix_7_1(gains):
    """Build 7.1 downmix matrix from gains dict."""
    C = C_TO_LR
    S = S_TO_LR
    R = REAR_TO_LR
    LFE = LFE_TO_LR
    
    matrix = {}
    matrix["FL"] = {"L": gains["left"], "R": 0.0}
    matrix["FR"] = {"L": 0.0, "R": gains["right"]}
    matrix["FC"] = {"L": gains["center"] * C, "R": gains["center"] * C}
    matrix["LFE"] = {"L": gains["lfe"] * LFE, "R": gains["lfe"] * LFE}
    matrix["SL"] = {"L": gains["side"] * S, "R": gains["side"] * S}
    matrix["SR"] = {"L": gains["side"] * S, "R": gains["side"] * S}
    matrix["BL"] = {"L": gains["rear"] * R, "R": gains["rear"] * R}
    matrix["BR"] = {"L": gains["rear"] * R, "R": gains["rear"] * R}
    return matrix

def compute_matrix_coeffs(matrix, gains_dict, channel_order, layout="5.1"):
    """Compute base downmix coefficients and final coefficients for each channel."""
    result = []
    key_map = CHANNEL_TO_GAIN_KEY_7_1 if layout == "7.1" else CHANNEL_TO_GAIN_KEY_5_1
    for ch in channel_order:
        coeff = matrix[ch]
        gain_key = key_map[ch]
        gain = gains_dict[gain_key]
        if gain != 0.0:
            base_L = coeff["L"] / gain
            base_R = coeff["R"] / gain
        else:
            base_L = 0.0
            base_R = 0.0
        result.append({
            "channel": ch,
            "configured_gain": gain,
            "base_downmix_coeff_L": round(base_L, 6),
            "base_downmix_coeff_R": round(base_R, 6),
            "final_L_coeff": round(coeff["L"], 6),
            "final_R_coeff": round(coeff["R"], 6)
        })
    return result

def build_matrix_for_report(label, layout, gains_dict):
    """Generate a report entry for a given configuration."""
    if layout == "7.1":
        matrix = build_matrix_7_1(gains_dict)
        channels = CHANNEL_NAMES_7_1
    else:
        matrix = build_matrix_5_1(gains_dict)
        channels = CHANNEL_NAMES_5_1
    
    coeffs = compute_matrix_coeffs(matrix, gains_dict, channels, layout)
    
    return {
        "label": label,
        "layout": layout,
        "coefficients": coeffs
    }

def main():
    report = {"matrix_contract": []}
    
    # 5.1 defaults
    report["matrix_contract"].append(build_matrix_for_report(
        "5.1 defaults", "5.1", DEFAULTS_5_1))
    
    # 7.1 defaults
    report["matrix_contract"].append(build_matrix_for_report(
        "7.1 defaults", "7.1", DEFAULTS_7_1))
    
    # 5.1 custom gains
    custom_51 = {"left": 0.8, "right": 1.2, "center": 0.5, "surround": 0.3, "lfe": 1.5}
    report["matrix_contract"].append(build_matrix_for_report(
        "5.1 custom gains", "5.1", custom_51))
    
    # 7.1 custom gains
    custom_71 = {"left": 0.5, "right": 1.5, "center": 0.5, 
                 "side": 0.9, "rear": 0.4, "lfe": 2.0}
    report["matrix_contract"].append(build_matrix_for_report(
        "7.1 custom gains", "7.1", custom_71))
    
    # 5.1 zero gains
    zero_51 = {k: 0.0 for k in DEFAULTS_5_1}
    report["matrix_contract"].append(build_matrix_for_report(
        "5.1 zero gains", "5.1", zero_51))
    
    # 7.1 zero gains
    zero_71 = {k: 0.0 for k in DEFAULTS_7_1}
    report["matrix_contract"].append(build_matrix_for_report(
        "7.1 zero gains", "7.1", zero_71))
    
    # 5.1 gains > 1.0
    high_51 = {"left": 2.0, "right": 1.5, "center": 3.0, "surround": 2.5, "lfe": 2.0}
    report["matrix_contract"].append(build_matrix_for_report(
        "5.1 gains > 1.0", "5.1", high_51))
    
    # 7.1 gains > 1.0
    high_71 = {"left": 1.5, "right": 2.0, "center": 1.5, 
               "side": 1.2, "rear": 1.3, "lfe": 2.0}
    report["matrix_contract"].append(build_matrix_for_report(
        "7.1 gains > 1.0", "7.1", high_71))
    
    # Matrix none (vanilla)
    report["matrix_contract"].append({
        "label": "matrix=none (vanilla)",
        "note": "All matrix_oba and matrix_cba = NONE (0). Vanilla linear downmix only."
    })
    
    print(json.dumps(report, indent=2))

if __name__ == "__main__":
    main()